From 203a3df5a239d644cf42b7bac03a268eb5babfc7 Mon Sep 17 00:00:00 2001
From: Alexander Larsson <alexl@redhat.com>
Date: Wed, 30 Aug 2017 11:38:09 +0900
Subject: [PATCH 11/33] Initial version of ibus portal

This adds a dbus service called org.freedesktop.portal.IBus on the
session bus. It is a very limited service that only implements
CreateInputContext and the InputContext interface (and Service.Destroy
for lifetime access).

It uses gdbus code generation for demarshalling the method calls which
means it will verify that all arguments have the right type.

Additionally all method calls to the input context object have to be
from the client that created it, so each client is isolated.

BUG=https://github.com/flatpak/flatpak/issues/675
R=Shawn.P.Huang@gmail.com

Review URL: https://codereview.appspot.com/326350043

Patch from Alexander Larsson <alexl@redhat.com>.
---
 Makefile.am                                   |   1 +
 configure.ac                                  |   5 +-
 portal/Makefile.am                            |  95 ++++
 portal/org.freedesktop.IBus.Portal.xml        | 132 +++++
 portal/org.freedesktop.portal.IBus.service.in |   3 +
 portal/portal.c                               | 698 ++++++++++++++++++++++++++
 src/ibusshare.h                               |  14 +
 7 files changed, 947 insertions(+), 1 deletion(-)
 create mode 100644 portal/Makefile.am
 create mode 100644 portal/org.freedesktop.IBus.Portal.xml
 create mode 100644 portal/org.freedesktop.portal.IBus.service.in
 create mode 100644 portal/portal.c

diff --git a/Makefile.am b/Makefile.am
index f703d4c6..c8e802da 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -51,6 +51,7 @@ SUBDIRS = \
 	util \
 	conf \
 	client \
+	portal \
 	data \
 	m4 \
 	po \
diff --git a/configure.ac b/configure.ac
index cb48ad4c..14556a3a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -153,7 +153,7 @@ PKG_CHECK_MODULES(GOBJECT2, [
     gobject-2.0 >= glib_required_version
 ])
 PKG_CHECK_MODULES(GIO2, [
-    gio-2.0 >= glib_required_version
+    gio-2.0 gio-unix-2.0 >= glib_required_version
 ])
 PKG_CHECK_MODULES(GTHREAD2, [
     gthread-2.0 >= glib_required_version
@@ -660,6 +660,8 @@ PKG_CHECK_MODULES(ISOCODES, [
 ISOCODES_PREFIX=`$PKG_CONFIG iso-codes --variable=prefix`
 AC_SUBST(ISOCODES_PREFIX)
 
+AC_SUBST([GDBUS_CODEGEN], [`$PKG_CONFIG --variable gdbus_codegen gio-2.0`])
+
 # OUTPUT files
 AC_CONFIG_FILES([ po/Makefile.in
 Makefile
@@ -674,6 +676,7 @@ src/Makefile
 src/ibusversion.h
 src/tests/Makefile
 bus/Makefile
+portal/Makefile
 engine/Makefile
 util/Makefile
 util/IMdkit/Makefile
diff --git a/portal/Makefile.am b/portal/Makefile.am
new file mode 100644
index 00000000..954fc591
--- /dev/null
+++ b/portal/Makefile.am
@@ -0,0 +1,95 @@
+# vim:set noet ts=4:
+#
+# ibus - The Input Bus
+#
+# Copyright (c) 2007-2013 Peng Huang <shawn.p.huang@gmail.com>
+# Copyright (c) 2007-2013 Red Hat, Inc.
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+# USA
+
+NULL =
+
+libibus = $(top_builddir)/src/libibus-@IBUS_API_VERSION@.la
+
+AM_CPPFLAGS = \
+	-I$(top_srcdir)/src   \
+	-I$(top_builddir)/src \
+	$(NULL)
+
+AM_CFLAGS = \
+	@GLIB2_CFLAGS@ \
+	@GIO2_CFLAGS@ \
+	@GTHREAD2_CFLAGS@ \
+	-DG_LOG_DOMAIN=\"IBUS\" \
+	-DPKGDATADIR=\"$(pkgdatadir)\" \
+	-DLIBEXECDIR=\"$(libexecdir)\" \
+	-DBINDIR=\"@bindir@\" \
+	-DIBUS_DISABLE_DEPRECATED \
+	$(NULL)
+AM_LDADD = \
+	@GOBJECT2_LIBS@ \
+	@GLIB2_LIBS@ \
+	@GIO2_LIBS@ \
+	@GTHREAD2_LIBS@ \
+	$(libibus) \
+	$(NULL)
+
+ibus_dbus_built_sources = ibus-portal-dbus.c ibus-portal-dbus.h
+BUILT_SOURCES = $(ibus_dbus_built_sources)
+
+libexec_PROGRAMS = ibus-portal
+ibus_portal_DEPENDENCIES = \
+	$(libibus) \
+	$(NULL)
+ibus_portal_SOURCES = \
+	portal.c \
+	$(ibus_dbus_built_sources) \
+	$(NULL)
+ibus_portal_CFLAGS = \
+	$(AM_CFLAGS) \
+	$(NULL)
+ibus_portal_LDADD = \
+	$(AM_LDADD) \
+	$(NULL)
+
+EXTRA_DIST = \
+	$(NULL)
+
+CLEANFILES = \
+	$(NULL)
+
+$(libibus):
+	$(MAKE) -C $(top_builddir)/src
+
+dbusservice_in_files = org.freedesktop.portal.IBus.service.in
+dbusservice_DATA = $(dbusservice_in_files:.service.in=.service)
+dbusservicedir=${datadir}/dbus-1/services
+
+org.freedesktop.portal.IBus.service: org.freedesktop.portal.IBus.service.in
+	$(AM_V_GEN) sed -e "s|\@libexecdir\@|$(libexecdir)|"  $< > $@.tmp && mv $@.tmp $@
+
+$(ibus_dbus_built_sources) : org.freedesktop.IBus.Portal.xml
+	$(AM_V_GEN) $(GDBUS_CODEGEN)                            \
+		--interface-prefix org.freedesktop.IBus.        \
+		--c-namespace IBusDbus                          \
+		--generate-c-code $(builddir)/ibus-portal-dbus  \
+		$^ \
+		$(NULL)
+
+EXTRA_DIST += $(dbusservice_in_files)
+CLEANFILES += $(dbusservice_DATA)
+
+-include $(top_srcdir)/git.mk
diff --git a/portal/org.freedesktop.IBus.Portal.xml b/portal/org.freedesktop.IBus.Portal.xml
new file mode 100644
index 00000000..afce4daa
--- /dev/null
+++ b/portal/org.freedesktop.IBus.Portal.xml
@@ -0,0 +1,132 @@
+<?xml version="1.0"?>
+<!--
+ Copyright (C) 2017 Red Hat, Inc.
+
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library. If not, see <http://www.gnu.org/licenses/>.
+
+ Author: Alexander Larsson <alexl@redhat.com>
+-->
+
+<node name="/" xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
+  <!--
+      org.freedesktop.IBus.Portal:
+      @short_description: Portal for ibus client access
+
+      This interface is a minimal interface to IBus that is safe to expose to
+      clients.
+  -->
+  <interface name="org.freedesktop.IBus.Portal">
+    <method name='CreateInputContext'>
+      <arg direction='in'  type='s' name='client_name' />
+      <arg direction='out' type='o' name='object_path' />
+    </method>
+  </interface>
+
+  <!-- This is a copy of the interface in inputcontext.c, they should be shared.
+       We want this for the code generator so that we can be sure we verify all
+       caller types, etc.
+  -->
+  <interface name='org.freedesktop.IBus.InputContext'>
+    <method name='ProcessKeyEvent'>
+      <arg direction='in'  type='u' name='keyval' />
+      <arg direction='in'  type='u' name='keycode' />
+      <arg direction='in'  type='u' name='state' />
+      <arg direction='out' type='b' name='handled' />
+    </method>
+    <method name='SetCursorLocation'>
+      <arg direction='in' type='i' name='x' />
+      <arg direction='in' type='i' name='y' />
+      <arg direction='in' type='i' name='w' />
+      <arg direction='in' type='i' name='h' />
+    </method>
+    <method name='SetCursorLocationRelative'>
+      <arg direction='in' type='i' name='x' />
+      <arg direction='in' type='i' name='y' />
+      <arg direction='in' type='i' name='w' />
+      <arg direction='in' type='i' name='h' />
+    </method>
+    <method name='ProcessHandWritingEvent'>
+      <arg direction='in' type='ad' name='coordinates' />
+    </method>
+    <method name='CancelHandWriting'>
+      <arg direction='in' type='u' name='n_strokes' />
+    </method>
+    <method name='FocusIn' />
+    <method name='FocusOut' />
+    <method name='Reset' />
+    <method name='SetCapabilities'>
+      <arg direction='in' type='u' name='caps' />
+    </method>
+    <method name='PropertyActivate'>
+      <arg direction='in' type='s' name='name' />
+      <arg direction='in' type='u' name='state' />
+    </method>
+    <method name='SetEngine'>
+      <arg direction='in' type='s' name='name' />
+    </method>
+    <method name='GetEngine'>
+      <arg direction='out' type='v' name='desc' />
+    </method>
+    <method name='SetSurroundingText'>
+      <arg direction='in' type='v' name='text' />
+      <arg direction='in' type='u' name='cursor_pos' />
+      <arg direction='in' type='u' name='anchor_pos' />
+    </method>
+
+    <signal name='CommitText'>
+      <arg type='v' name='text' />
+    </signal>
+    <signal name='ForwardKeyEvent'>
+      <arg type='u' name='keyval' />
+      <arg type='u' name='keycode' />
+      <arg type='u' name='state' />
+    </signal>
+    <signal name='UpdatePreeditText'>
+      <arg type='v' name='text' />
+      <arg type='u' name='cursor_pos' />
+      <arg type='b' name='visible' />
+    </signal>
+    <signal name='ShowPreeditText'/>
+    <signal name='HidePreeditText'/>
+    <signal name='UpdateAuxiliaryText'>
+      <arg type='v' name='text' />
+      <arg type='b' name='visible' />
+    </signal>
+    <signal name='ShowAuxiliaryText'/>
+    <signal name='HideAuxiliaryText'/>
+    <signal name='UpdateLookupTable'>
+      <arg type='v' name='table' />
+      <arg type='b' name='visible' />
+    </signal>
+    <signal name='ShowLookupTable'/>
+    <signal name='HideLookupTable'/>
+    <signal name='PageUpLookupTable'/>
+    <signal name='PageDownLookupTable'/>
+    <signal name='CursorUpLookupTable'/>
+    <signal name='CursorDownLookupTable'/>
+    <signal name='RegisterProperties'>
+      <arg type='v' name='props' />
+    </signal>
+    <signal name='UpdateProperty'>
+      <arg type='v' name='prop' />
+    </signal>
+
+    <property name='ContentType' type='(uu)' access='write' />
+  </interface>
+
+  <interface name='org.freedesktop.IBus.Service'>
+    <method name='Destroy' />
+  </interface>
+
+</node>
diff --git a/portal/org.freedesktop.portal.IBus.service.in b/portal/org.freedesktop.portal.IBus.service.in
new file mode 100644
index 00000000..47ae9ffc
--- /dev/null
+++ b/portal/org.freedesktop.portal.IBus.service.in
@@ -0,0 +1,3 @@
+[D-BUS Service]
+Name=org.freedesktop.portal.IBus
+Exec=@libexecdir@/ibus-portal
diff --git a/portal/portal.c b/portal/portal.c
new file mode 100644
index 00000000..0415f996
--- /dev/null
+++ b/portal/portal.c
@@ -0,0 +1,698 @@
+/* -*- mode: C; c-basic-offset: 4; indent-tabs-mode: nil; -*- */
+/* vim:set et sts=4: */
+/* ibus - The Input Bus
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include <fcntl.h>
+#include <glib.h>
+#include <gio/gio.h>
+#include <ibus.h>
+#include <locale.h>
+#include <pwd.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "ibus-portal-dbus.h"
+
+typedef struct _IBusPortal IBusPortal;
+typedef struct _IBusPortalClass IBusPortalClass;
+typedef struct _IBusPortalContext IBusPortalContext;
+typedef struct _IBusPortalContextClass IBusPortalContextClass;
+
+struct _IBusPortalContext
+{
+    IBusDbusInputContextSkeleton parent_instance;
+    IBusInputContext *context;
+    guint id;
+    char *owner;
+    char *object_path;
+    IBusDbusService *service;
+};
+
+struct _IBusPortalContextClass
+{
+    IBusDbusInputContextSkeletonClass parent_class;
+};
+
+struct _IBusPortal
+{
+    IBusDbusPortalSkeleton parent_instance;
+
+};
+
+struct _IBusPortalClass
+{
+    IBusDbusPortalSkeletonClass parent_class;
+};
+
+enum
+{
+    PROP_CONTENT_TYPE = 1,
+    N_PROPERTIES
+};
+
+static GMainLoop *loop = NULL;
+static IBusBus *ibus_bus;
+static IBusPortal *ibus_portal = NULL;
+static gboolean opt_verbose;
+static gboolean opt_replace;
+
+static GList *all_contexts = NULL;
+
+static guint next_context_id;
+
+GType ibus_portal_context_get_type (void) G_GNUC_CONST;
+static void ibus_portal_context_iface_init (IBusDbusInputContextIface *iface);
+
+static void portal_context_g_signal (GDBusProxy        *proxy,
+                                     const gchar       *sender_name,
+                                     const gchar       *signal_name,
+                                     GVariant          *parameters,
+                                     IBusPortalContext *portal_context);
+
+G_DEFINE_TYPE_WITH_CODE (IBusPortalContext,
+                         ibus_portal_context,
+                         IBUS_DBUS_TYPE_INPUT_CONTEXT_SKELETON,
+                         G_IMPLEMENT_INTERFACE (IBUS_DBUS_TYPE_INPUT_CONTEXT,
+                                 ibus_portal_context_iface_init));
+
+static void
+_forward_method_cb (GObject *source_object,
+                    GAsyncResult *res,
+                    gpointer user_data)
+{
+    GDBusMethodInvocation *invocation = user_data;
+    IBusPortalContext *portal_context =
+            (IBusPortalContext *) g_dbus_method_invocation_get_user_data (
+                    invocation);
+    IBusEngineDesc *desc;
+    GError *error = NULL;
+
+    GVariant *variant = g_dbus_proxy_call_finish ((GDBusProxy *) source_object,
+                                                  res, &error);
+    if (variant == NULL) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        return;
+    }
+
+    g_dbus_method_invocation_return_value (invocation, variant);
+}
+
+static gboolean
+_forward_method (IBusDbusInputContext  *object,
+                 GDBusMethodInvocation *invocation)
+{
+    IBusPortalContext *portal_context = (IBusPortalContext *)object;
+    GDBusMessage *message = g_dbus_method_invocation_get_message (invocation);
+
+    g_dbus_proxy_call (G_DBUS_PROXY (portal_context->context),
+                       g_dbus_method_invocation_get_method_name (invocation),
+                       g_dbus_message_get_body (message),
+                       G_DBUS_CALL_FLAGS_NONE,
+                       -1,
+                       NULL, /* cancellable */
+                       _forward_method_cb, invocation);
+    return TRUE;
+}
+
+static gboolean
+ibus_dbus_context_cancel_hand_writing (IBusDbusInputContext  *object,
+                                       GDBusMethodInvocation *invocation,
+                                       guint                  arg_n_strokes)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_focus_in (IBusDbusInputContext  *object,
+                            GDBusMethodInvocation *invocation)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_focus_out (IBusDbusInputContext  *object,
+                             GDBusMethodInvocation *invocation)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_get_engine (IBusDbusInputContext  *object,
+                              GDBusMethodInvocation *invocation)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_process_hand_writing_event (IBusDbusInputContext  *object,
+                                              GDBusMethodInvocation *invocation,
+                                              GVariant
+                                                               *arg_coordinates)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_process_key_event (IBusDbusInputContext  *object,
+                                     GDBusMethodInvocation *invocation,
+                                     guint                  arg_keyval,
+                                     guint                  arg_keycode,
+                                     guint                  arg_state)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_property_activate (IBusDbusInputContext  *object,
+                                     GDBusMethodInvocation *invocation,
+                                     const gchar           *arg_name,
+                                     guint                  arg_state)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_reset (IBusDbusInputContext  *object,
+                         GDBusMethodInvocation *invocation)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_set_capabilities (IBusDbusInputContext  *object,
+                                    GDBusMethodInvocation *invocation,
+                                    guint                  arg_caps)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_set_cursor_location (IBusDbusInputContext  *object,
+                                       GDBusMethodInvocation *invocation,
+                                       gint                   arg_x,
+                                       gint                   arg_y,
+                                       gint                   arg_w,
+                                       gint                   arg_h)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_set_cursor_location_relative (IBusDbusInputContext  *object,
+                                                GDBusMethodInvocation
+                                                                    *invocation,
+                                                gint                   arg_x,
+                                                gint                   arg_y,
+                                                gint                   arg_w,
+                                                gint                   arg_h)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_set_engine (IBusDbusInputContext  *object,
+                              GDBusMethodInvocation *invocation,
+                              const gchar           *arg_name)
+{
+    return _forward_method (object, invocation);
+}
+
+static gboolean
+ibus_dbus_context_set_surrounding_text (IBusDbusInputContext  *object,
+                                        GDBusMethodInvocation *invocation,
+                                        GVariant              *arg_text,
+                                        guint                  arg_cursor_pos,
+                                        guint                  arg_anchor_pos)
+{
+    return _forward_method (object, invocation);
+}
+
+static void
+ibus_portal_context_iface_init (IBusDbusInputContextIface *iface)
+{
+    iface->handle_cancel_hand_writing = ibus_dbus_context_cancel_hand_writing;
+    iface->handle_focus_in = ibus_dbus_context_focus_in;
+    iface->handle_focus_out = ibus_dbus_context_focus_out;
+    iface->handle_get_engine = ibus_dbus_context_get_engine;
+    iface->handle_process_hand_writing_event =
+            ibus_dbus_context_process_hand_writing_event;
+    iface->handle_process_key_event = ibus_dbus_context_process_key_event;
+    iface->handle_property_activate = ibus_dbus_context_property_activate;
+    iface->handle_reset = ibus_dbus_context_reset;
+    iface->handle_set_capabilities = ibus_dbus_context_set_capabilities;
+    iface->handle_set_cursor_location = ibus_dbus_context_set_cursor_location;
+    iface->handle_set_cursor_location_relative =
+            ibus_dbus_context_set_cursor_location_relative;
+    iface->handle_set_engine = ibus_dbus_context_set_engine;
+    iface->handle_set_surrounding_text = ibus_dbus_context_set_surrounding_text;
+}
+
+static void
+ibus_portal_context_init (IBusPortalContext *portal_context)
+{
+}
+
+static void
+ibus_portal_context_finalize (GObject *object)
+{
+    IBusPortalContext *portal_context = (IBusPortalContext *)object;
+
+    all_contexts = g_list_remove (all_contexts, portal_context);
+
+    g_dbus_interface_skeleton_unexport (
+            G_DBUS_INTERFACE_SKELETON (portal_context->service));
+    g_dbus_interface_skeleton_unexport (
+            G_DBUS_INTERFACE_SKELETON (portal_context));
+
+    g_free (portal_context->owner);
+    g_free (portal_context->object_path);
+    g_object_unref (portal_context->service);
+
+    g_signal_handlers_disconnect_by_func (
+            portal_context->context,
+            G_CALLBACK(portal_context_g_signal),
+            portal_context);
+    g_object_unref (portal_context->context);
+
+    G_OBJECT_CLASS (ibus_portal_context_parent_class)->finalize (object);
+}
+
+static void
+ibus_portal_context_set_property (IBusPortalContext *portal_context,
+                                  guint              prop_id,
+                                  const GValue      *value,
+                                  GParamSpec        *pspec)
+{
+    switch (prop_id) {
+    case PROP_CONTENT_TYPE:
+        g_dbus_proxy_call (G_DBUS_PROXY (portal_context->context),
+                           "org.freedesktop.DBus.Properties.Set",
+                           g_variant_new ("(ssv)",
+                                          IBUS_INTERFACE_INPUT_CONTEXT,
+                                          "ContentType",
+                                          g_value_get_variant (value)),
+                           G_DBUS_CALL_FLAGS_NONE,
+                           -1,
+                           NULL, /* cancellable */
+                           NULL, /* callback */
+                           NULL  /* user_data */
+                           );
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (portal_context, prop_id, pspec);
+    }
+}
+
+static void
+ibus_portal_context_get_property (IBusPortalContext *portal_context,
+                                  guint              prop_id,
+                                  GValue            *value,
+                                  GParamSpec        *pspec)
+{
+    switch (prop_id) {
+    case PROP_CONTENT_TYPE:
+        g_warning ("No support for setting content type");
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (portal_context, prop_id, pspec);
+    }
+}
+
+static gboolean
+ibus_portal_context_g_authorize_method (GDBusInterfaceSkeleton *interface,
+                                        GDBusMethodInvocation  *invocation)
+{
+    IBusPortalContext *portal_context = (IBusPortalContext *)interface;
+
+    if (g_strcmp0 (g_dbus_method_invocation_get_sender (invocation),
+                   portal_context->owner) == 0) {
+        return TRUE;
+    }
+
+    g_dbus_method_invocation_return_error (invocation,
+                                           G_DBUS_ERROR,
+                                           G_DBUS_ERROR_FAILED,
+                                           "Access denied");
+    return FALSE;
+}
+
+
+static void
+ibus_portal_context_class_init (IBusPortalContextClass *klass)
+{
+    GObjectClass *gobject_class;
+    GDBusInterfaceSkeletonClass *skeleton_class;
+
+    gobject_class = G_OBJECT_CLASS (klass);
+    gobject_class->finalize  = ibus_portal_context_finalize;
+    gobject_class->set_property  =
+            (GObjectSetPropertyFunc) ibus_portal_context_set_property;
+    gobject_class->get_property  =
+            (GObjectGetPropertyFunc) ibus_portal_context_get_property;
+
+    skeleton_class = G_DBUS_INTERFACE_SKELETON_CLASS(klass);
+    skeleton_class->g_authorize_method = ibus_portal_context_g_authorize_method;
+
+    ibus_dbus_input_context_override_properties (gobject_class,
+                                                 PROP_CONTENT_TYPE);
+}
+
+static void
+portal_context_g_signal (GDBusProxy        *proxy,
+                         const gchar       *sender_name,
+                         const gchar       *signal_name,
+                         GVariant          *parameters,
+                         IBusPortalContext *portal_context)
+{
+    GError *error = NULL;
+    GDBusConnection *connection;
+
+    if (g_strcmp0 (sender_name, IBUS_SERVICE_IBUS) != 0)
+        return;
+
+    connection = g_dbus_interface_skeleton_get_connection (
+            G_DBUS_INTERFACE_SKELETON (portal_context));
+    if (!g_dbus_connection_emit_signal (connection,
+                                        portal_context->owner,
+                                        portal_context->object_path,
+                                        IBUS_INTERFACE_INPUT_CONTEXT,
+                                        signal_name,
+                                        parameters,
+                                        &error)) {
+        g_warning ("Unable to emit signal %s: %s", signal_name, error->message);
+        g_error_free (error);
+    }
+
+    g_signal_stop_emission_by_name (proxy, "g-signal");
+}
+
+static gboolean
+ibus_portal_context_handle_destroy (IBusDbusService       *object,
+                                    GDBusMethodInvocation *invocation,
+                                    IBusPortalContext     *portal_context)
+{
+    g_object_unref (portal_context);
+}
+
+static IBusPortalContext *
+ibus_portal_context_new (IBusInputContext *context,
+                         GDBusConnection  *connection,
+                         const char       *owner,
+                         GError          **error)
+{
+    IBusPortalContext *portal_context =
+            g_object_new (ibus_portal_context_get_type (), NULL);
+
+    g_signal_connect (context,
+                      "g-signal",
+                      G_CALLBACK(portal_context_g_signal),
+                      portal_context);
+
+    portal_context->id = ++next_context_id;
+    portal_context->context = g_object_ref (context);
+    portal_context->owner = g_strdup (owner);
+    portal_context->object_path =
+            g_strdup_printf (IBUS_PATH_INPUT_CONTEXT, portal_context->id);
+    portal_context->service = ibus_dbus_service_skeleton_new ();
+
+    g_signal_connect (portal_context->service,
+                      "handle-destroy",
+                      G_CALLBACK (ibus_portal_context_handle_destroy),
+                      portal_context);
+
+    if (!g_dbus_interface_skeleton_export (
+                G_DBUS_INTERFACE_SKELETON (portal_context->service),
+                connection, portal_context->object_path,
+                error) ||
+        !g_dbus_interface_skeleton_export (
+                G_DBUS_INTERFACE_SKELETON (portal_context),
+                connection, portal_context->object_path,
+                error)) {
+        g_object_unref (portal_context);
+        return NULL;
+    }
+
+    all_contexts = g_list_prepend (all_contexts, portal_context);
+
+    return portal_context;
+}
+
+GType ibus_portal_get_type (void) G_GNUC_CONST;
+static void ibus_portal_iface_init (IBusDbusPortalIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (IBusPortal, ibus_portal,
+                         IBUS_DBUS_TYPE_PORTAL_SKELETON,
+                         G_IMPLEMENT_INTERFACE (IBUS_DBUS_TYPE_PORTAL,
+                                                ibus_portal_iface_init));
+
+
+static void
+create_input_context_done (IBusBus               *bus,
+                           GAsyncResult          *res,
+                           GDBusMethodInvocation *invocation)
+{
+    GError *error = NULL;
+    IBusInputContext *context;
+    IBusPortalContext *portal_context;
+    char *object_path;
+
+    context = ibus_bus_create_input_context_async_finish (ibus_bus,
+                                                          res,
+                                                          &error);
+    if (context == NULL) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        return;
+    }
+
+    portal_context = ibus_portal_context_new (
+            context,
+            g_dbus_method_invocation_get_connection (invocation),
+            g_dbus_method_invocation_get_sender (invocation),
+            &error);
+    g_object_unref (context);
+
+    if (portal_context == NULL) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        g_object_unref (portal_context);
+        return;
+    }
+
+    ibus_dbus_portal_complete_create_input_context (
+            IBUS_DBUS_PORTAL(ibus_portal),
+            invocation, portal_context->object_path);
+}
+
+static gboolean
+ibus_portal_handle_create_input_context (IBusDbusPortal        *object,
+                                         GDBusMethodInvocation *invocation,
+                                         const gchar           *arg_client_name)
+{
+    ibus_bus_create_input_context_async (
+            ibus_bus,
+            arg_client_name, -1,
+            NULL,
+            (GAsyncReadyCallback)create_input_context_done,
+            invocation);
+    return TRUE;
+}
+
+static void
+ibus_portal_iface_init (IBusDbusPortalIface *iface)
+{
+    iface->handle_create_input_context =
+            ibus_portal_handle_create_input_context;
+}
+
+static void
+ibus_portal_init (IBusPortal *portal)
+{
+}
+
+static void
+ibus_portal_class_init (IBusPortalClass *klass)
+{
+}
+
+
+static void
+show_version_and_quit (void)
+{
+    g_print ("%s - Version %s\n", g_get_application_name (), VERSION);
+    exit (EXIT_SUCCESS);
+}
+
+static const GOptionEntry entries[] =
+{
+    { "version",   'V', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
+      show_version_and_quit, "Show the application's version.", NULL },
+    { "verbose",   'v', 0, G_OPTION_ARG_NONE, 
+      &opt_verbose,   "verbose.", NULL },
+    { "replace",   'r', 0, G_OPTION_ARG_NONE,
+      &opt_replace,   "Replace.", NULL },
+    { NULL },
+};
+
+static void
+on_bus_acquired (GDBusConnection *connection,
+                 const gchar     *name,
+                 gpointer         user_data)
+{
+    GError *error = NULL;
+
+    ibus_portal = g_object_new (ibus_portal_get_type (), NULL);
+
+    if (!g_dbus_interface_skeleton_export (
+                G_DBUS_INTERFACE_SKELETON (ibus_portal),
+                connection,
+                IBUS_PATH_IBUS,
+                &error)) {
+        g_warning ("Error exporting portal: %s", error->message);
+        g_error_free (error);
+        return;
+    }
+}
+
+static void
+on_name_acquired (GDBusConnection *connection,
+                  const gchar     *name,
+                  gpointer         user_data)
+{
+}
+
+static void
+on_name_lost (GDBusConnection *connection,
+              const gchar     *name,
+              gpointer         user_data)
+{
+    g_main_loop_quit (loop);
+}
+
+static void
+name_owner_changed (GDBusConnection *connection,
+                    const gchar     *sender_name,
+                    const gchar     *object_path,
+                    const gchar     *interface_name,
+                    const gchar     *signal_name,
+                    GVariant        *parameters,
+                    gpointer         user_data)
+{
+  const char *name, *from, *to;
+
+  g_variant_get (parameters, "(sss)", &name, &from, &to);
+
+  if (name[0] == ':' &&
+      g_strcmp0 (name, from) == 0 &&
+      g_strcmp0 (to, "") == 0)
+    {
+        GList *l, *next;
+        /* Client disconnected, free any input contexts it may have */
+        for (l = all_contexts; l != NULL; l = next) {
+            IBusPortalContext *portal_context = l->data;
+            next = l->next;
+
+            if (g_strcmp0 (portal_context->owner, name) == 0) {
+                g_object_unref (portal_context);
+            }
+        }
+    }
+}
+
+static void
+_bus_disconnected_cb (IBusBus            *ibusbus)
+{
+    g_main_loop_quit (loop);
+}
+
+gint
+main (gint argc, gchar **argv)
+{
+    GDBusConnection *session_bus = NULL;
+    guint owner_id;
+
+    setlocale (LC_ALL, "");
+
+    GOptionContext *context = g_option_context_new ("- ibus daemon");
+    g_option_context_add_main_entries (context, entries, "ibus-daemon");
+
+    GError *error = NULL;
+    if (!g_option_context_parse (context, &argc, &argv, &error)) {
+        g_printerr ("Option parsing failed: %s\n", error->message);
+        g_error_free (error);
+        exit (-1);
+    }
+
+    /* Avoid even loading gvfs to avoid accidental confusion */
+    g_setenv ("GIO_USE_VFS", "local", TRUE);
+
+    ibus_init ();
+
+    ibus_set_log_handler (opt_verbose);
+
+    ibus_bus = ibus_bus_new ();
+    if (!ibus_bus_is_connected (ibus_bus)) {
+        g_printerr ("Not connected to the ibus bus\n");
+        exit (1);
+    }
+
+    g_signal_connect (ibus_bus, "disconnected",
+                      G_CALLBACK (_bus_disconnected_cb), NULL);
+
+    loop = g_main_loop_new (NULL, FALSE);
+
+    session_bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
+    if (session_bus == NULL) {
+        g_printerr ("No session bus: %s", error->message);
+        exit (-1);
+    }
+
+    g_dbus_connection_signal_subscribe (session_bus,
+                                        "org.freedesktop.DBus",
+                                        "org.freedesktop.DBus",
+                                        "NameOwnerChanged",
+                                        "/org/freedesktop/DBus",
+                                        NULL,
+                                        G_DBUS_SIGNAL_FLAGS_NONE,
+                                        name_owner_changed,
+                                        NULL, NULL);
+
+    owner_id = g_bus_own_name (G_BUS_TYPE_SESSION,
+                               IBUS_SERVICE_PORTAL,
+                               G_BUS_NAME_OWNER_FLAGS_ALLOW_REPLACEMENT |
+                               (opt_replace ? G_BUS_NAME_OWNER_FLAGS_REPLACE
+                                            : 0),
+                               on_bus_acquired,
+                               on_name_acquired,
+                               on_name_lost,
+                               NULL,
+                               NULL);
+
+    g_main_loop_run (loop);
+
+    g_bus_unown_name (owner_id);
+    g_main_loop_unref (loop);
+
+    return 0;
+}
diff --git a/src/ibusshare.h b/src/ibusshare.h
index bca477c0..f3e2011e 100644
--- a/src/ibusshare.h
+++ b/src/ibusshare.h
@@ -51,6 +51,13 @@
  */
 #define IBUS_SERVICE_IBUS       "org.freedesktop.IBus"
 
+/**
+ * IBUS_SERVICE_PORTAL:
+ *
+ * Address of IBus portalservice.
+ */
+#define IBUS_SERVICE_PORTAL     "org.freedesktop.portal.IBus"
+
 /**
  * IBUS_SERVICE_PANEL:
  *
@@ -121,6 +128,13 @@
  */
 #define IBUS_INTERFACE_IBUS     "org.freedesktop.IBus"
 
+/**
+ * IBUS_INTERFACE_PORTAL:
+ *
+ * D-Bus interface for IBus portal.
+ */
+#define IBUS_INTERFACE_PORTAL   "org.freedesktop.IBus.Portal"
+
 /**
  * IBUS_INTERFACE_INPUT_CONTEXT:
  *
-- 
2.15.1

